[
    {
        "url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/26",
        "repository_url": "https://api.github.com/repos/MisaOgura/flashtorch",
        "labels_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/26/labels{/name}",
        "comments_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/26/comments",
        "events_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/26/events",
        "html_url": "https://github.com/MisaOgura/flashtorch/issues/26",
        "id": 558407634,
        "node_id": "MDU6SXNzdWU1NTg0MDc2MzQ=",
        "number": 26,
        "title": "How to handle multi label multiclass?",
        "user": {
            "login": "rrags",
            "id": 7228934,
            "node_id": "MDQ6VXNlcjcyMjg5MzQ=",
            "avatar_url": "https://avatars3.githubusercontent.com/u/7228934?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/rrags",
            "html_url": "https://github.com/rrags",
            "followers_url": "https://api.github.com/users/rrags/followers",
            "following_url": "https://api.github.com/users/rrags/following{/other_user}",
            "gists_url": "https://api.github.com/users/rrags/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/rrags/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/rrags/subscriptions",
            "organizations_url": "https://api.github.com/users/rrags/orgs",
            "repos_url": "https://api.github.com/users/rrags/repos",
            "events_url": "https://api.github.com/users/rrags/events{/privacy}",
            "received_events_url": "https://api.github.com/users/rrags/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2020-01-31T22:29:31Z",
        "updated_at": "2020-01-31T22:30:02Z",
        "closed_at": null,
        "author_association": "NONE",
        "body": "My network does binary classifcation for detection of 11 different classes. E.G. It predicts if or if not there are apples, oranges, pineapple, and pears in the image I give it. So the output is a binary vector of length 11. \r\n\r\nCan I use this project with out modifying it? \r\nI have checked and top_label will always be 9, so what I put for target_label will be ignored. I get the error\r\n\r\n`The predicted class index 9 does notequal the target class index 3. Calculatingthe gradient w.r.t. the predicted class.\r\n  'the gradient w.r.t. the predicted class.'`\r\n\r\nSo in my example, if the network outputs a binary vector of length 4 corresponding to whether or not apple, orange, pineapple, and pear are in the image, how can I make it so that when I set target = 3 the code will show the gradient corresponding to the task of detecting pineapples? \r\n\r\nAlso, I am using modified ResNet-18 (transfer learned). "
    },
    {
        "url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/24",
        "repository_url": "https://api.github.com/repos/MisaOgura/flashtorch",
        "labels_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/24/labels{/name}",
        "comments_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/24/comments",
        "events_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/24/events",
        "html_url": "https://github.com/MisaOgura/flashtorch/issues/24",
        "id": 541963103,
        "node_id": "MDU6SXNzdWU1NDE5NjMxMDM=",
        "number": 24,
        "title": "How can I use FlashTorch to see what the EfficientNet learn?",
        "user": {
            "login": "ChenXiao61",
            "id": 39661761,
            "node_id": "MDQ6VXNlcjM5NjYxNzYx",
            "avatar_url": "https://avatars1.githubusercontent.com/u/39661761?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/ChenXiao61",
            "html_url": "https://github.com/ChenXiao61",
            "followers_url": "https://api.github.com/users/ChenXiao61/followers",
            "following_url": "https://api.github.com/users/ChenXiao61/following{/other_user}",
            "gists_url": "https://api.github.com/users/ChenXiao61/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/ChenXiao61/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/ChenXiao61/subscriptions",
            "organizations_url": "https://api.github.com/users/ChenXiao61/orgs",
            "repos_url": "https://api.github.com/users/ChenXiao61/repos",
            "events_url": "https://api.github.com/users/ChenXiao61/events{/privacy}",
            "received_events_url": "https://api.github.com/users/ChenXiao61/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2019-12-24T02:28:21Z",
        "updated_at": "2020-01-02T10:41:17Z",
        "closed_at": null,
        "author_association": "NONE",
        "body": "**Describe the bug**\r\nA clear and concise description of what the bug is.\r\n\r\n**To Reproduce**\r\nSteps to reproduce the behavior:\r\n1. Go to '...'\r\n2. Click on '....'\r\n3. Scroll down to '....'\r\n4. See error\r\n\r\n**Expected behavior**\r\nA clear and concise description of what you expected to happen.\r\n\r\n**Screenshots**\r\nIf applicable, add screenshots to help explain your problem.\r\n\r\n**Environment (please complete the following information):**\r\n - OS: [e.g. iOS]\r\n - Python version: [e.g. 3.6]\r\n - FlashTorch version: [e.g. 0.1.0]\r\n\r\n**Additional context**\r\nAdd any other context about the problem here.\r\n"
    },
    {
        "url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/23",
        "repository_url": "https://api.github.com/repos/MisaOgura/flashtorch",
        "labels_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/23/labels{/name}",
        "comments_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/23/comments",
        "events_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/23/events",
        "html_url": "https://github.com/MisaOgura/flashtorch/issues/23",
        "id": 531534143,
        "node_id": "MDU6SXNzdWU1MzE1MzQxNDM=",
        "number": 23,
        "title": "[BUG] Backprop.calculate_gradients gives rise to \" AttributeError: 'NoneType' object has no attribute 'shape' \" error",
        "user": {
            "login": "narbhar",
            "id": 47373022,
            "node_id": "MDQ6VXNlcjQ3MzczMDIy",
            "avatar_url": "https://avatars2.githubusercontent.com/u/47373022?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/narbhar",
            "html_url": "https://github.com/narbhar",
            "followers_url": "https://api.github.com/users/narbhar/followers",
            "following_url": "https://api.github.com/users/narbhar/following{/other_user}",
            "gists_url": "https://api.github.com/users/narbhar/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/narbhar/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/narbhar/subscriptions",
            "organizations_url": "https://api.github.com/users/narbhar/orgs",
            "repos_url": "https://api.github.com/users/narbhar/repos",
            "events_url": "https://api.github.com/users/narbhar/events{/privacy}",
            "received_events_url": "https://api.github.com/users/narbhar/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 1285118171,
                "node_id": "MDU6TGFiZWwxMjg1MTE4MTcx",
                "url": "https://api.github.com/repos/MisaOgura/flashtorch/labels/bug",
                "name": "bug",
                "color": "d73a4a",
                "default": true,
                "description": "Something isn't working"
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 1,
        "created_at": "2019-12-02T20:57:22Z",
        "updated_at": "2020-02-01T18:54:35Z",
        "closed_at": null,
        "author_association": "NONE",
        "body": "**Describe the bug**\r\nWhen testing the visualization tool using a custom pre-trained CNN that performs Binary Classification, the error \"AttributeError: 'NoneType' object has no attribute 'shape'\" crops up when trying to visualize an image that from the dataset used to train the model. \r\n\r\n**To Reproduce**\r\nSteps to reproduce the behavior:\r\nLoad image using custom transformation:\r\n`transform = transforms.Compose([\r\n    transforms.RandomHorizontalFlip(),\r\n    transforms.RandomRotation(20),\r\n    transforms.ToTensor(),\r\n    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))\r\n    ])\r\n`\r\n that pertains to the CNN model developed.\r\nLoad the image and apply the transformation:\r\n`from flashtorch.utils import load_image`\r\n`from flashtorch.saliency import Backprop`\r\n`backprop=Backprop(n)`\r\n`image =load_image(\"img0042.png\")`\r\n`i = transform(image)`\r\n`i = torch.unsqueeze(i,0)`\r\n`target_class=0`\r\n`gradients=backprop.calculate_gradients(i,0)`\r\n(general summary of the code)\r\n\r\n\r\n\r\n**Error message**\r\n---------------------------------------------------------------------------\r\nAttributeError                            Traceback (most recent call last)\r\n<ipython-input-36-18e9aeeac78d> in <module>\r\n      1 from torch.autograd import Variable\r\n----> 2 gradients=backprop.calculate_gradients(i,0)\r\n\r\n~\\Anaconda3\\lib\\site-packages\\flashtorch\\saliency\\backprop.py in calculate_gradients(self, input_, target_class, take_max, guided, use_gpu)\r\n    118         # Calculate gradients of the target class output w.r.t. input_\r\n    119 \r\n--> 120         output.backward(gradient=target)\r\n    121 \r\n    122         # Detach the gradients from the graph and move to cpu\r\n\r\n~\\Anaconda3\\lib\\site-packages\\torch\\tensor.py in backward(self, gradient, retain_graph, create_graph)\r\n    164                 products. Defaults to ``False``.\r\n    165         \"\"\"\r\n--> 166         torch.autograd.backward(self, gradient, retain_graph, create_graph)\r\n    167 \r\n    168     def register_hook(self, hook):\r\n\r\n~\\Anaconda3\\lib\\site-packages\\torch\\autograd\\__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)\r\n     97     Variable._execution_engine.run_backward(\r\n     98         tensors, grad_tensors, retain_graph, create_graph,\r\n---> 99         allow_unreachable=True)  # allow_unreachable flag\r\n    100 \r\n    101 \r\n\r\n~\\Anaconda3\\lib\\site-packages\\flashtorch\\saliency\\backprop.py in _record_gradients(module, grad_in, grad_out)\r\n    212     def _register_conv_hook(self):\r\n    213         def _record_gradients(module, grad_in, grad_out):\r\n--> 214             if self.gradients.shape == grad_in[0].shape:\r\n    215                 self.gradients = grad_in[0]\r\n    216 \r\n\r\nAttributeError: 'NoneType' object has no attribute 'shape'\r\n\r\n**Environment (please complete the following information):**\r\n - OS: [Windows 10 1909 version: 18363.476]\r\n - Python version: [3.7..4]\r\n - FlashTorch version: [0.1.1]\r\n"
    },
    {
        "url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/17",
        "repository_url": "https://api.github.com/repos/MisaOgura/flashtorch",
        "labels_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/17/labels{/name}",
        "comments_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/17/comments",
        "events_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/17/events",
        "html_url": "https://github.com/MisaOgura/flashtorch/issues/17",
        "id": 508149857,
        "node_id": "MDU6SXNzdWU1MDgxNDk4NTc=",
        "number": 17,
        "title": "backprop.visualize does not display anything",
        "user": {
            "login": "goldblum",
            "id": 442416,
            "node_id": "MDQ6VXNlcjQ0MjQxNg==",
            "avatar_url": "https://avatars2.githubusercontent.com/u/442416?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/goldblum",
            "html_url": "https://github.com/goldblum",
            "followers_url": "https://api.github.com/users/goldblum/followers",
            "following_url": "https://api.github.com/users/goldblum/following{/other_user}",
            "gists_url": "https://api.github.com/users/goldblum/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/goldblum/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/goldblum/subscriptions",
            "organizations_url": "https://api.github.com/users/goldblum/orgs",
            "repos_url": "https://api.github.com/users/goldblum/repos",
            "events_url": "https://api.github.com/users/goldblum/events{/privacy}",
            "received_events_url": "https://api.github.com/users/goldblum/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 1285118171,
                "node_id": "MDU6TGFiZWwxMjg1MTE4MTcx",
                "url": "https://api.github.com/repos/MisaOgura/flashtorch/labels/bug",
                "name": "bug",
                "color": "d73a4a",
                "default": true,
                "description": "Something isn't working"
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 2,
        "created_at": "2019-10-16T23:23:44Z",
        "updated_at": "2019-12-06T16:51:23Z",
        "closed_at": null,
        "author_association": "NONE",
        "body": "I followed the tutorial on the README page, and the visualizations are not displaying.  I am using a pretrained model, and I use a tensor with requires_grad=True as the image.  My code doesn't run in to any errors.  My code is below:\r\n\r\nbackprop = Backprop(net)\r\nbackprop.visualize(image, target_class, guided = True)"
    },
    {
        "url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/16",
        "repository_url": "https://api.github.com/repos/MisaOgura/flashtorch",
        "labels_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/16/labels{/name}",
        "comments_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/16/comments",
        "events_url": "https://api.github.com/repos/MisaOgura/flashtorch/issues/16/events",
        "html_url": "https://github.com/MisaOgura/flashtorch/issues/16",
        "id": 507950785,
        "node_id": "MDU6SXNzdWU1MDc5NTA3ODU=",
        "number": 16,
        "title": "Gradient Bug in backprop.visualize",
        "user": {
            "login": "goldblum",
            "id": 442416,
            "node_id": "MDQ6VXNlcjQ0MjQxNg==",
            "avatar_url": "https://avatars2.githubusercontent.com/u/442416?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/goldblum",
            "html_url": "https://github.com/goldblum",
            "followers_url": "https://api.github.com/users/goldblum/followers",
            "following_url": "https://api.github.com/users/goldblum/following{/other_user}",
            "gists_url": "https://api.github.com/users/goldblum/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/goldblum/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/goldblum/subscriptions",
            "organizations_url": "https://api.github.com/users/goldblum/orgs",
            "repos_url": "https://api.github.com/users/goldblum/repos",
            "events_url": "https://api.github.com/users/goldblum/events{/privacy}",
            "received_events_url": "https://api.github.com/users/goldblum/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 1285118171,
                "node_id": "MDU6TGFiZWwxMjg1MTE4MTcx",
                "url": "https://api.github.com/repos/MisaOgura/flashtorch/labels/bug",
                "name": "bug",
                "color": "d73a4a",
                "default": true,
                "description": "Something isn't working"
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 6,
        "created_at": "2019-10-16T16:08:29Z",
        "updated_at": "2019-12-10T14:54:05Z",
        "closed_at": null,
        "author_association": "NONE",
        "body": "I tried generating a saliency map using the following code:\r\n\r\n    backprop = Backprop(net)\r\n    target_class = 4\r\n    backprop.visualize(image, target_class, guided=True)\r\n\r\nMy net is a standard PyTorch neural network, and the image is a (1,3,84,84) image.  However, I get the following error:\r\n\r\n File \"visualizations.py\", line 89, in test\r\n    backprop.visualize(image, target_class, guided=True)\r\n  File \"/home/user/.local/lib/python3.6/site-packages/flashtorch/saliency/backprop.py\", line 168, in visualize\r\n    guided=guided)\r\n  File \"/home/user/.local/lib/python3.6/site-packages/flashtorch/saliency/backprop.py\", line 120, in calculate_gradients\r\n    output.backward(gradient=target)\r\n  File \"/home/user/.local/lib/python3.6/site-packages/torch/tensor.py\", line 150, in backward\r\n    torch.autograd.backward(self, gradient, retain_graph, create_graph)\r\n  File \"/home/user/.local/lib/python3.6/site-packages/torch/autograd/__init__.py\", line 99, in backward\r\n    allow_unreachable=True)  # allow_unreachable flag\r\n  File \"/home/user/.local/lib/python3.6/site-packages/flashtorch/saliency/backprop.py\", line 214, in _record_gradients\r\n    if self.gradients.shape == grad_in[0].shape:\r\nAttributeError: 'NoneType' object has no attribute 'shape'\r\n\r\nDo I need to turn the image into a parameter with requires_grad = True?\r\nIs this usually done by the load_image function?  I found that this function was re-sizing my image to 224x224.  Thanks for your help!"
    }
]